### DISCUSSION TITLE
Tailwind CSS 3 version on iOS 12 or older have some problem

### DISCUSSION THREAD

**Discussion by sarukomine** [tailwindlabs/tailwindcss#6814] [Help] [answered]
After we upgraded to 3 version for our website, and we just found some styles couldn't be displayed normally on Safari with iOS 12.

In 3 version,
e.g. `background-color: rgb(244 0 0/var(--tw-bg-opacity));`
This style is not working for iOS 12, but iOS 13 or latest

In 2 version,
e.g. `background-color: rgba(244, 0, 0, var(--tw-bg-opacity));`

Referring to Apple Report (https://developer.apple.com/support/app-store/), only 7% users are using iOS 12 or older.

**Comment by adamwathan** [comment, marked as answer]
Yeah Tailwind 3 targets Safari 14/15, and we don’t make any effort to support older versions.

**Reply by Ikem-Emmanuel** [reply]
This issues still persists with higher version like 14/15...Example: button styles are completely different on safari but okay on chrome..

Please if u have a suggestion kindly share. Thanks

**Reply by rishabkapadia** [reply]
yes. facing the same issue. @Ikem-Emmanuel did you find any solution?
tried this as well. https://stackoverflow.com/a/71324360/2419875

**Reply by li-nd** [reply]
It helped to solve most of the problems in my case

```
  // tailwind.config.js
  module.exports = {
    corePlugins: {
        backdropOpacity: false,
        backgroundOpacity: false,
        borderOpacity: false,
        divideOpacity: false,
        ringOpacity: false,
        textOpacity: false
    }
  }
```

**Comment by CornellVanDerStraaten** [comment]
This is also an issue for me in Safari 14, background styles not working. This has not been answered yet.

**Comment by jonnitto** [comment]
You could include https://github.com/dmarchena/postcss-color-rgb to get the old syntax back

**Reply by rishabkapadia** [reply]
this package is vulnerable to denial of service. pls avoid it.

**Comment by kristjanmar** [comment]
I just upgraded to Tailwind 3 this morning, then later decided to test my site on BrowserStack and the colors are all missing on Safari 12.

Unfortunately 1.4% of my users are using Safari 12 so it needs to have at least minimal support and not look completely broken.

So I added background-color: #XXXXXX; to my css file for the most crucial elements that need to have a color.

This included overlays or dropdowns or sticky elements that need to have a white background or else they are transparent and useless. I also did this for links, button backgrounds and such.

```css
.dropdown {
   @apply absolute right-0 z-50 mt-2 min-w-[160px] origin-top-right rounded-md bg-white py-1 shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none;
   background-color: #ffffff;
}
```

I just converted the Tailwind color to HEX code, so it looks identical from what I can see. And the site doesn't look completely broken on Safari 12 anymore.

Is there any downside to this approach?

**Comment by zoos** [comment]
translate 
```css
rgb(244 0 0/var(--tw-bg-opacity))
```
to
```css
rgba(244, 0, 0, var(--tw-bg-opacity))
```

bash script
```sh
cssfile=$(ls index*.css)
cat $cssfile\
|sed -r 's/rgb\(([0-9]+)\s+([0-9]+)\s+([0-9]+)\/(var\([a-z-]+\))\)/rgba(\1,\2,\3,\4)/g'\
|tee $cssfile  >/dev/null
```

**Comment by icebearlabs** [comment]
As of February 2023 this is still an issue in current iOS versions

**Comment by medzhidov** [comment]
Meh, Tailwind Team, unsupported <=iphone 12 is awful solution... 

But anyway, all with this problem can just use my custom plugin:
https://github.com/medzhidov/rgb-to-rgba-postcss

Hope this will help.

@sarukomine @icebearlabs @zoos @kristjanmar @CornellVanDerStraaten @adamwathan @VSKut @rishabkapadia

**Reply by adamwathan** [reply]
Hey! iOS 14 is supported on every iPhone from the iPhone 6S and above, not the iPhone 12. And commaless rgb syntax is supported as far back as iOS 12. If you need to support legacy browsers I would recommend the PostCSS plugin above or use Tailwind CSS v2.

**Reply by zoos** [reply]
and what is the objective need to use the new syntax?  the color is displayed the same way in both the old and the new syntax the same way.  only the number of supporting devices is less.  what is the advantage of the new syntax?

**Comment by savancodexive** [comment]
**the `bg-opacity-70` class does not exist. If `bg-opacity-70` is a custom class, make sure it is defined within a `@layer` directive.**

**Reply by medzhidov** [reply]
Where did you find "bg-opacity-70" in this thread?
